164796886 User Registration Mail Verification Feature

Simplemart17 opened the pull request:
#### Description
When user signup, there is need to verify their account. This feature write function to send welcome mail and a verification link to confirm the email address.
#### Type of change
- New feature (non-breaking change which add functionality)
#### How Has This Been Tested?
- Integration testing
- Unit testing
#### Checklist:
- My code follows the style guidelines of this project
- My changes generate no new warning
- New and existing unit test pass locally with my changes
#### PT
[#164796886](https://www.pivotaltracker.com/story/show/164796886)

Simplemart17 committed:
164796886-feature(emailVerification): add function for user verification

- add function for user to verify their email address after registration

[Delivers #164796886]

Simplemart17 committed:
164796886-feature(emailVerification): implement email verification

- set up email verification

[Deliver #164796886]

Simplemart17 committed:
164796886-feature(emailVerification): set up email verification

- add function to send email after successfult signup

[Delivers #164796886]

Simplemart17 committed:
164796886-feature(emailVerification): function for user to receieve email after signup

- add function for user to receieve verification link after signup
- add route for verifying user after receiving the mail

[Deliver #164796886]

Simplemart17 committed:
164796886-feature(emailVerification): add test for verification feature

- add unit test and integration test for email verification feature

[Delivers #164796886]

chikeozulumba commented on src/utils/email.js:
```diff
@@ -0,0 +1,16 @@
+};
+
+
+// const msg = {
```
@Simplemart17  Great work on this feature. I could not wrap my head around why these comments should be on file, would you consider resolving it by removing it or applying any other method as you see fit. Thanks

Simplemart17 replied:
It has been resolved. Thanks

ssanusi commented on src/controllers/confirmation.js:
```diff
@@ -0,0 +1,38 @@
+      const { secretToken } = req.params;
+      const user = await User.findOne({
+        where: { secretToken }
+      },);
```
@Simplemart17 please attend to houndCi

Temmyogunbo commented on src/controllers/authentication/user.js:
```diff
@@ -9,52 +12,57 @@ import { responseFormat, errorResponseFormat } from '../../utils/index';
 
+    let url = `http://${process.env.HOST}:${process.env.PORT}/api/v1/verify`;
+
+    if (process.env.NODE_ENV === 'production') { url = `https://${process.env.HOST}:${process.env.PORT}/api/v1/verify`; }
```
Let us learn to separate code and put them where they should belong. This should not be here.

Simplemart17 replied:
It has been resolved. Thanks

Temmyogunbo commented on src/controllers/authentication/user.js:
```diff
@@ -9,52 +12,57 @@ import { responseFormat, errorResponseFormat } from '../../utils/index';
+        <p>Hi ${fullName},</p>
+        <p>Welcome to Authors Haven, a place to be inspired! Your account was successfully created.</p>
+        <p>Please click this <a href=${verificationUrl}>link</a> to confirm your account.</p>
+      </div>`;
```
This should not here.

Temmyogunbo commented on src/controllers/authentication/user.js:
```diff
@@ -9,52 +12,57 @@ import { responseFormat, errorResponseFormat } from '../../utils/index';
+    let user = await User.findOne({ where: { id, verificationToken } });
+    if (!user) { return res.status(403).json(errorResponseFormat({ status: 'fail', message: 'Invalid token supplied, kindly reauthenticate!', })); }
+    user = await user.update({ verified: true, verificationToken: null });
+    if (user) { return res.status(202).json(responseFormat({ status: 'success', data: 'Your account was successfully verified!', })); }
```
Think of line 62 and 64, it is difficult to reason. Let us improve on the names

Simplemart17 replied:
I have made it more readable. Thanks

Simplemart17 committed:
164796886-feature(emailVerification): implement feedback

- modify email verification function

[Delivers #164796886]

Simplemart17 committed:
164796886-feature(emailVerification): update branch

- update branch to resolve merge conflict

[Delivers #164796886]

Simplemart17 committed:
164796886-feature(emailVerification): modify verification redirect

- redirect user after successful email verification
- fixed failed travis build

[Delivers #164796886]
